src/retryOnEmpty.ts

```diff
@@ -0,0 +1,109 @@
+import {
+  createAsyncMiddleware,
+  JsonRpcMiddleware,
+  PendingJsonRpcResponse,
+} from 'json-rpc-engine';
+import clone from 'clone';
+import pify from 'pify';
+import {
+  Block,
+  blockTagParamIndex,
+  SafeEventEmitterProvider,
+} from './cache-utils';
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
+const BlockTracker = require('eth-block-tracker');
+
+//
+// RetryOnEmptyMiddleware will retry any request with an empty response that has
+// a numbered block reference at or lower than the blockTracker's latest block.
+// Its useful for dealing with load-balanced ethereum JSON RPC
+// nodes that are not always in sync with each other.
+//
+
+export = createRetryOnEmptyMiddleware;
+
+// empty values used to determine if a request should be retried
+// `<nil>` comes from https://github.com/ethereum/go-ethereum/issues/16925
+const emptyValues: (string|null|undefined)[] = [undefined, null, '\u003cnil\u003e'];
+
+interface RetryOnEmptyMiddlewareOptions{
+  provider?: SafeEventEmitterProvider;
+  blockTracker?: typeof BlockTracker;
+}
+
+function createRetryOnEmptyMiddleware(
+  opts: RetryOnEmptyMiddlewareOptions = {}
```

Comment:
As noted elsewhere, `opts` should be destructured here, if possible. We can keep the default for now.